### ISSUE TITLE
Some thoughts on experiments

### ISSUE THREAD

**Issue by krezelj** [krezelj/academia#162]
We will of course discuss experiments in more details during the next meeting but I've decided to create an issue for my thoughts and possibly for some pre-discussion. I've already created some issues but I'll keep them open for now. These are
- #139 
- #146 
- #156 
- #161 

## Saving all stats
This one is probably obvious but during the experiments we should really just save all stats unless an experiment was a failure (because it was not correct or took too much time etc.). What I mean is that we shouldn't delete saved stats after creating visualisations. As I've said, it's probably obvious but because it's so important I still wanted to point it out.

## Experiment reviews
I think that before starting a specific experiment we should have them undergo a review process similar to PR reviews (maybe even make it a PR into `experiments` directory to make use of Github's reviewing tools). Since they do take a lot of time and if something is wrong that only shows at the end it's not as easy as "rerunning it in the morning" we should really let everyone take a look to make sure the entire setup is correct and share potential issues with the experiment itself.

That being said we will probably discuss specific experiments during our meetings (what we want to experiment and why). But the code and configuration should be reviewed as well.

## Experiment READMEs
If we decide to go with the `experiment` directory and make subdirectories for each experiment it might be beneficial to create a README for each experiment. Not only for us, but also for our advisor, reviewing committee and possibly other interested parties.
They will also be helpful when writing our thesis and as such they should be rather formal and technical.

## `epsilon` and `epsilon_decay` during experiments
(Copied from #139)
When experimenting with epsilon-greedy agents (such as DQN) we need to remember to correctly set starting epsilon and more importantly epsilon_decay. If it's too small the epsilon will quickly fall to 0 (or min_epsilon) making it impossible to explore.

## Max Steps
(Copied from #146)
Nocurriculum task should use `max_steps` set to the `total_step_count` from the experiment.
From my initial experiments `3*mean_total_steps` seems to work well. However I think that to make experiments more consistent we should first run all curriculum runs and only then calculate the mean and run the no curriculum runs. This includes scenarios where we run the experiments in batches i.e. we shouldn't run one batch consisting of n curriculum runs and m nocurriculum runs and then repeat the batch but instead we should first run all curriculum runs and only then do the nocurriculum. This sadly applies to my already gathered stats for `DoorKey` since it didn't do it this way. But since it didn't take long to run that experiment I think it's okay if we repeat it just so that it's consistent with the rest of the experiments.

## `BridgeBuilding` experiments
(Copied from #156)
I've already closed an issue related to this (https://github.com/krezelj/wut-curriculum-learning/issues/29 ) as not planned however on second thought it might actually be an interesting idea to perform this experiment. Especially since we have an environment over which we have total control.

The experiment would simply compare two curricula with the same configuration but run on different BridgeBuilding configurations. One with sparse reward and one with dense reward. Furthermore, my initial experiments indicate that RIVER_WIDTH=3 might be a bit too challenging for the agents. RIVER_WIDTH=2 seems a bit better. This means that there would be a few more changes to the BridgeBuilding environment but nothing significant.

The changes would include

adding a reward_density parameter ("dense" or "sparse")
changing default river_width to 2.
adding a river_width parameter (now allowed difficulties would include numbers between 0 and river_width)
My idea for the dense reward is as follows:

If the agent drowns or the episode terminates because of max_steps reached the reward is 0
However a reward is added for the furthest placed bridge segment and it's value is 1/river_width * max_bridge_x (max_bridge_x is relative to the left bank not left edge of the map). However max_bridge_x is capped at river_width so that the agent doesn't try to dismantle the bridge to place one of the boulders on the right bank and obtain a higher than intended reward.
If the agent reaches the goal the rewards is sparse_goal_reward + 1 (where sparse_goal_reward is the reward it would get in an environment with reward_density="sparse" set). +1 is added so that reaching the goal is still always beneficial.
I think it might be beneficial to our "investigation of impact of curriculum learning" to see how the impact differs in two essentially the same environments but with different reward functions. Especially since curriculum learning is supposed to mitigate the negative effects of sparse rewards so we would expect it having a larger impact in environments with a sparse reward function.

## `epsilon` reset value in curriculum
(Copied from #161)
had a thought why DQN might perform poorly in Szymon's initial experiments.

PPO after training on an easy task tries to exploit what it has learned and when it doesn't work it tries to do that less and less slowly trying to explore more and more and then it stumbles on a good tactic again and works towards that.

With DQN we are currently reseting the epsilon to 1 on each new task and let it slowly fall down which makes the agent prioritise exploration over exploitation which in turn might make it forget some things it has learned. This would explain the perceived fall in performence Szymon has pointed out.

It might make more sense to reset the value to something much smaller like 0.2 or even 0.1. In fact, an experiment that checks how the reset value affects curriculum performance could be one of the experiments we perform as it's not an obvious thing to consider (at least to me). We could compare curriculum trajectories for 3 different reset values and compare results.

Either way we probably need a way to control this epsilon reset value inside LearningTask. Probably the best solution in the long run would be to somehow include task callbacks -- functions that would be called after a specific task has finished. This would let us have total control over any changes to the agent we want to make after each task (https://github.com/krezelj/wut-curriculum-learning/issues/21 :) ). But this is obviously an overkill for this specific issue especially because it would complicate things with saving/loading tasks and further delay experiments. However I think some action is needed as DQN curriculum trajectories looked very bleak in comparison to PPO trajectories and I don't believe it's because of PPO superiority over DQN (at least not entirely) especially since I've read that on-policy algorithms should perform worse on sparse reward tasks.

## Jupyter Notebooks
Another thing to consider is whether we want to run our experiments using jupyer notebooks
**Pros**
- Easy and clean seperation of experiment setup, experiment running, stats loading and stats visualisations in one file using different cells
- Easy for other people to see the executed jupyter notebook (e.g. with plots) using github

**Cons**
- jupyer notebooks work horribly with git
